src/main/kotlin/org/jitsi/nlj/transform/node/incoming/RemoteBandwidthEstimator.kt

```diff
@@ -0,0 +1,105 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jitsi.nlj.transform.node.incoming
+
+import org.jitsi.nlj.PacketInfo
+import org.jitsi.nlj.rtp.RtpExtensionType.ABS_SEND_TIME
+import org.jitsi.nlj.stats.NodeStatsBlock
+import org.jitsi.nlj.transform.node.ObserverNode
+import org.jitsi.nlj.util.ReadOnlyStreamInformationStore
+import org.jitsi.nlj.util.createChildLogger
+import org.jitsi.nlj.util.observableWhenChanged
+import org.jitsi.rtp.rtcp.rtcpfb.payload_specific_fb.RtcpFbRembPacket
+import org.jitsi.rtp.rtcp.rtcpfb.payload_specific_fb.RtcpFbRembPacketBuilder
+import org.jitsi.rtp.rtp.RtpPacket
+import org.jitsi.rtp.rtp.header_extensions.AbsSendTimeHeaderExtension
+import org.jitsi.utils.LRUCache
+import org.jitsi.utils.logging.DiagnosticContext
+import org.jitsi.utils.logging2.Logger
+import org.jitsi_modified.impl.neomedia.rtp.remotebitrateestimator.RemoteBitrateEstimatorAbsSendTime
+import java.time.Clock
+
+/**
+ * Estimates the available bandwidth for the incoming stream using the abs-send-time extension.
+ */
+class RemoteBandwidthEstimator(
+    streamInformationStore: ReadOnlyStreamInformationStore,
+    parentLogger: Logger,
+    diagnosticContext: DiagnosticContext = DiagnosticContext(),
+    private val clock: Clock = Clock.systemUTC()
+) : ObserverNode("Remote Bandwidth Estimator") {
+    private val logger = parentLogger.createChildLogger(RemoteBandwidthEstimator::class)
+    /**
+     * The remote bandwidth estimation is enabled when REMB support is signaled, but TCC is not signaled.
+     */
+    private var enabled: Boolean by observableWhenChanged(false) {
+        _, _, newValue -> logger.debug { "Setting enabled=$newValue." }
+    }
+    private var astExtId: Int? = null
+    private var rbe: RemoteBitrateEstimatorAbsSendTime = RemoteBitrateEstimatorAbsSendTime(diagnosticContext, logger)
+    private val ssrcs = LRUCache<Long, Int>(MAX_SSRCS, true)
+    private var numRembsCreated = 0
+    private var numPacketsWithoutAbsSendTime = 0
+
+    init {
+        streamInformationStore.onRtpExtensionMapping(ABS_SEND_TIME) {
+            astExtId = it
+            logger.debug { "Setting abs-send-time extension ID to $astExtId" }
+        }
+        streamInformationStore.onRtpPayloadTypesChanged {
+            enabled = streamInformationStore.supportsRemb && !streamInformationStore.supportsTcc
+        }
+    }
+
+    companion object {
+        private const val MAX_SSRCS: Int = 8
+    }
+
+    override fun observe(packetInfo: PacketInfo) {
+        if (!enabled) return
+        val rtpPacket: RtpPacket = packetInfo.packet as? RtpPacket ?: return
```

Comment:
I think we use `val rtpPacket = packetInfo.packetAs<RtpPacket>()` in these places since it should be impossible to fail (since we're in the RTP path here), and we'd want to get an exception during testing or whatever should it actually fail so we can look into it.

Reply:
Also, once we do that I think we can put this inside the `astExtId?.let` block.

Reply:
👍